### PR TITLE
Export data processor widget

### PR DESCRIPTION
**Pull request by AntonPiccardoSelg**
This is a second attempt to export the data processor widget.

Please see the original PR for more information: #18922
The original issue is #14732

**To test:**

No issue number here

No release notes.

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)? Is it well structured with small focussed classes/methods/functions?
- [ ] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] How do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant documentation been added/updated?
- [ ] Is user-facing documentation written in a user-friendly manner?
- [ ] Has developer documentation been updated if required?

- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by AntonPiccardoSelg** [comment]
@PranavBahuguna  The issues on the `DataProcessorWidget` PR have finally been resolved and it is ready for final testing The python 3 tests are failing because of the general PowderDiffraction issue). To build it locally you might have to clean the `CustomInterfaces` target. `sip` apparently does not play well with incremental builds.

**Comment by AntonPiccardoSelg** [comment]
Given the timeline and since this is a non-user facing change that will only be used in features in `3.11`. I makes sense to push this to `3.11`.

File: scripts/Interface/ui/dataprocessorinterface/data_processor_gui.py [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+
+    def setup_layout(self):
+        """
+        Do further setup that could not be done in the designer.
+        So far only two menus have been added, we need to add the processing table manually.
+        """
+
+        # The white list (mandatory)
+        # Defines the number of columns, their names and the algorithm properties linked to them
+        # Metdho 'addElement' adds a column to the widget
```
Comment (PranavBahuguna) [inline scripts/Interface/ui/dataprocessorinterface/data_processor_gui.py]: Typo - "Method"

File: MantidQt/MantidWidgets/src/DataProcessorUI/GenericDataProcessorPresenter.cpp [hunk 2]
```diff
@@ -672,28 +731,34 @@ GenericDataProcessorPresenter::reduceRow(const std::vector<std::string> &data) {
...
+                                              : "";
+
+      auto optionsMap = parseKeyValueString(globalOptionsForColumn);
       auto runWS =
           prepareRunWorkspace(preProcessValue, preprocessor, optionsMap);
       alg->setProperty(propertyName, runWS->getName());
     } else {
       // No pre-processing needed
-      alg->setPropertyValue(propertyName, preProcessValue);
+      // No pre-processing needed
```
Comment (PranavBahuguna) [inline MantidQt/MantidWidgets/src/DataProcessorUI/GenericDataProcessorPresenter.cpp]: Can get rid of this comment.

File: MantidQt/MantidWidgets/src/DataProcessorUI/GenericDataProcessorPresenter.cpp [hunk 3]
```diff
@@ -672,28 +731,34 @@ GenericDataProcessorPresenter::reduceRow(const std::vector<std::string> &data) {
...
+      auto propertyValue = data.at(i);
+      if (!propertyValue.empty())
+        alg->setPropertyValue(propertyName, propertyValue);
+      // alg->setPropertyValue(propertyName, globalOptions);
```
Comment (PranavBahuguna) [inline MantidQt/MantidWidgets/src/DataProcessorUI/GenericDataProcessorPresenter.cpp]: Do we need this if it is commented out?

File: docs/source/interfaces/DataProcessorWidget.rst [hunk 4]
```diff
@@ -0,0 +1,751 @@
...
+															  'SecondTransmissionWorkspace,'
+															  'MomentumTransferMin,'
+															  'MomentumTransferMax,'
+															  'MomentumTransferStep,'
+															  'ScaleFactor,'
+															  'OutputWorkspaceBinned,'
+															  'OutputWorkspace,'
+															  'OutputWorkspaceWavelength,')
+
+The only effect of the blaklist is on the *Options* column, not in the reduction. This column uses a *HintingLineEdit* (a MantidWidget)
```
Comment (PranavBahuguna) [inline docs/source/interfaces/DataProcessorWidget.rst]: Typo - 'blacklist'

File: docs/source/interfaces/DataProcessorWidget.rst [hunk 5]
```diff
@@ -0,0 +1,751 @@
...
+| DataProcessorCommand                |                                                          |
++-------------------------------------+----------------------------------------------------------+
+| DataProcessorCommandBase            |                                                          |
++-------------------------------------+----------------------------------------------------------+
+| DataProcessor***Command             |                                                          |
++-------------------------------------+----------------------------------------------------------+
+| DataProcessorCommandAdapter         |                                                          |
++-------------------------------------+----------------------------------------------------------+
+| DataProcessorGenerateNotebook       |                                                          |
++-------------------------------------+----------------------------------------------------------+
```
Comment (PranavBahuguna) [inline docs/source/interfaces/DataProcessorWidget.rst]: There are no descriptions of these classes. Should there be descriptions or has it been left blank on purpose?

**Review by PranavBahuguna — CHANGES_REQUESTED** [review]
A major problem I found is when attempting to process any rows, the interface concatenates all property values set as a single string and attempts to use it as a value for the `Filename` property, throwing an error. I suspect this is related to your changes since this was not an error that existed when I had previously reviewed this with Raquel.

There are also a few minor issues in documentation and some typos that need addressing.

**Comment by AntonPiccardoSelg** [comment]
Once the async stuff is in master I will address these issues

**Comment by PranavBahuguna** [comment]
New code seems good, probably best to just remove the class descriptions for now. Good to go once that is done and checks pass.

**Comment by PranavBahuguna** [comment]
:shipit: once checks pass

**Comment by PranavBahuguna** [comment]
:shipit:

**Review by PranavBahuguna — APPROVED** [review]

**Comment by ianbush** [comment]
I assume the inclusion of `MantidQt/MantidWidgets/src/DataProcessorUI/GenericDataProcessorPresenter.cpp.orig` is a mistake?

**Pull request merged**